File: backend/pkg/data/models/products/powersupply.go
package products

import (
	"BetterPC_2.0/pkg/data/models/products/general"
	generalResponses "BetterPC_2.0/pkg/data/models/products/general/responses"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"strconv"
)

type PowerSupply struct {
	ID          primitive.ObjectID `bson:"_id,omitempty" json:"id,omitempty"`
	General     general.General    `bson:"general" json:"general"`
	FormFactor  string             `bson:"form_factor" json:"form_factor,omitempty"`
	OutputPower int                `bson:"output_power" json:"output_power,omitempty"`
	Connectors  Connectors         `bson:"connectors" json:"connectors"`
	Modules     bool               `bson:"modules" json:"modules,omitempty"`
	MbPower     int                `bson:"mb_power" json:"mb_power,omitempty"`
	CpuPower    CpuPower           `bson:"cpu_power" json:"cpu_power,omitempty"`
}

type Connectors struct {
	Sata  int   `bson:"sata,omitempty" json:"sata,omitempty"`
	Molex int   `bson:"molex,omitempty" json:"molex,omitempty"`
	PciE  []int `bson:"pci_e,omitempty" json:"pci_e,omitempty"`
}

type CpuPower struct {
	Amount int   `bson:"amount" json:"amount,omitempty"`
	Type   []int `bson:"type" json:"type,omitempty"`
}

func (powerSupply *PowerSupply) GetModel() string {
	return powerSupply.General.Model
}

func (powerSupply *PowerSupply) GetStock() int {
	return powerSupply.General.Amount
}

func (powerSupply *PowerSupply) GetImage() string {
	return powerSupply.General.Image
}

func (powerSupply *PowerSupply) SetImage(imageName string) {
	powerSupply.General.Image = imageName
}

func (powerSupply *PowerSupply) Standardize() generalResponses.StandardizedProductData {
	var product generalResponses.StandardizedProductData
	product.ProductHeader.ID = powerSupply.ID.Hex()
	product.ProductHeader.ProductType = "powersupply"
	product.Name = powerSupply.General.Model
	product.General = powerSupply.General
	product.Description = "Form Factor: " + powerSupply.FormFactor + ", Output Power: " +
		strconv.Itoa(powerSupply.OutputPower) + "W, Modular: " + strconv.FormatBool(powerSupply.Modules)
	return product
}

func (powerSupply *PowerSupply) CalculateFinalPrice() int {
	return powerSupply.General.GetFinalPrice()
}

File: frontend/better_pc/src/modules/products/models/powersupply.ts
import {BaseProduct} from "@/modules/products/models/product_common"; // Общий интерфейс General

// Описание структуры подключения (SATA, Molex, PCI-E)
export interface Connectors {
    sata?: number; // Количество портов SATA
    molex?: number; // Количество разъемов Molex
    pciE?: number[]; // Список слотов PCI-E (например, [6, 8] для двух разъемов PCI-E 6-pin и 8-pin)
}

// Описание структуры питания процессора
export interface CpuPower {
    amount?: number; // Количество разъемов питания процессора
    type?: number[]; // Типы разъемов (например, [4, 8] для разъемов 4-pin и 8-pin)
}

export interface PowerSupply extends BaseProduct {
    formFactor?: string; // Форм-фактор блока питания (например, ATX, SFX)
    outputPower?: number; // Выходная мощность блока питания (в ваттах)
    connectors: Connectors; // Разъемы блока питания (SATA, Molex, PCI-E)
    modules?: boolean; // Наличие модульных кабелей
    mbPower?: number; // Питание материнской платы (в пинах)
    cpuPower?: CpuPower; // Питание процессора, включая количество и тип разъемов
}
